docs/notebooks/JAX_pytrees.ipynb

```diff
@@ -0,0 +1,323 @@
+        "\n",
+        "We refer to a recursive structured value whose leaves are basic types as a `pytree`. When JAX flattens a pytree it will produce a list of leaves and a `treedef` object that encodes the structure of the original value. The `treedef` can then be used to construct a matching structured value after transforming the leaves. Pytrees are tree-like, rather than DAG-like or graph-like, in that we handle them assuming referential transparency and that they can't contain reference cycles.   \n",
+        "\n",
+        "Here is a simple example:\n"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "X8DlAmOMmufl",
+        "colab_type": "code",
+        "outputId": "2e7821f3-7d0d-48c2-e9d3-6e19cef07255",
+        "colab": {
+          "base_uri": "https://localhost:8080/",
+          "height": 85
+        }
+      },
+      "source": [
+        "from jax.tree_util import tree_flatten, tree_unflatten, register_pytree_node\n",
+        "from jax import numpy as np\n",
+        "\n",
+        "# The structured value to be transformed\n",
+        "value_structured = [1., (2., 3.)]\n",
+        "\n",
+        "# The leaves in value_flat correspond to the `*` markers in value_tree\n",
+        "value_flat, value_tree = tree_flatten(value_structured)\n",
+        "print(\"value_flat={}\\nvalue_tree={}\".format(value_flat, value_tree))\n",
+        "\n",
+        "# Transform the flt value list using an element-wise numeric transformer\n",
+        "transformed_flat = list(map(lambda v: v * 2., value_flat))\n",
+        "print(\"transformed_flat={}\".format(transformed_flat))\n",
+        "\n",
+        "# Reconstruct the structured output, using the original \n",
+        "transformed_structured = tree_unflatten(value_tree, transformed_flat)\n",
+        "print(\"transformed_structured={}\", transformed_structured)"
+      ],
+      "execution_count": 1,
+      "outputs": [
+        {
+          "output_type": "stream",
+          "text": [
+            "value_flat=[1.0, 2.0, 3.0]\n",
+            "value_tree=PyTreeDef(list, [*,PyTreeDef(tuple, [*,*])])\n",
+            "transformed_flat=[2.0, 4.0, 6.0]\n",
+            "transformed_structured={} [2.0, (4.0, 6.0)]\n"
+          ],
+          "name": "stdout"
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "sgUJpiXSsRSi",
+        "colab_type": "text"
+      },
+      "source": [
+        "Pytrees containers can be lists, tuples, dicts, namedtuple. Numeric and ndarray values are treated as leaves:"
```

Comment:
The comment about numeric and ndarray values is a bit misleading (indeed, see the following section.)

Nit: `NamedTuple`. We also support a few more things via custom registration (e.g., `OrderedDict`).

`None` is also currently considered an interior node.

Reply:
I have added OrderedDict and None to the list of internal nodes. I changed the comment about the numeric values:

"Other types of values, including numeric and ndarray values, are treated as leaves"

I hope that's what you meant.